liblangutil/Scanner.cpp

```diff
@@ -271,12 +273,74 @@ bool Scanner::skipWhitespaceExceptUnicodeLinebreak()
 	return sourcePos() != startPosition;
 }
 
+// A little helper class for counting depth of RLO/LRO/PDF uses.
+struct Scanner::UnicodeDirectionOverrideProcessor
+{
+	Scanner& scanner;
+	int rtlOverrideDepth = 0;
+
+	/// Tries to scan for an RLO/LRO/PDF and keeps track of script writing direction override depth.
+	///
+	/// @returns a 2-tuple indicating whether or not successfully some script writing direction
+	///          override has been consumed from the input and an error code in case
+	///          the input's lexical parser state is invalid and this error should be reported
+	///          to the user.
+	pair<bool, ScannerError> operator()()
+	{
+		// U+202D (LRO - Left-to-Right Override)
+		// U+202E (RLO - Right-to-Left Override)
+		if (
+			scanner.tryScanByteSequence("\xE2\x80\xAD") ||
+			scanner.tryScanByteSequence("\xE2\x80\xAE")
+		)
+		{
+			rtlOverrideDepth++;
+			return pair{true, ScannerError::NoError};
+		}
+		else if (scanner.tryScanByteSequence("\xE2\x80\xAC")) // U+202C (PDF - Pop Directional Formatting)
+		{
+			rtlOverrideDepth--;
+			if (rtlOverrideDepth < 0)
+				return pair{true, ScannerError::DirectionalOverrideUnderflowInComment};
+			else
+				return pair{true, ScannerError::NoError};
+		}
+		else
+			return pair{false, ScannerError::NoError};
+	}
+
+	ScannerError error() const noexcept
+	{
+		if (rtlOverrideDepth < 0)
+			return ScannerError::DirectionalOverrideUnderflowInComment;
+		else if (rtlOverrideDepth > 0)
+			return ScannerError::DirectionalOverrideMismatchInComment;
+		else
+			return ScannerError::NoError;
+	}
+};
+
 Token Scanner::skipSingleLineComment()
 {
+	UnicodeDirectionOverrideProcessor unicodeDirectionalOverride{*this};
+
 	// Line terminator is not part of the comment. If it is a
 	// non-ascii line terminator, it will result in a parser error.
 	while (!isUnicodeLinebreak())
-		if (!advance()) break;
+	{
+		auto const [processed, errorCode] = unicodeDirectionalOverride();
+		if (processed && errorCode != ScannerError::NoError)
+			return setError(errorCode);
+		else if (!processed)
```

Comment:
I was wondering how this doesn't enter an infinite loop when `processed` is `true` and there's no error but then I realized it's because `unicodeDirectionalOverride()` internally calls `tryScanByteSequence()` which does the advancing.

I think that a function with a proper name would be more self-explanatory than `operator()`.

Reply:
Renaming `processed` to `advanced` or `streamAdvanced` might also help.

Reply:
we have quite some `...scan...` functions that are advancing. I just adapted to that style. I can rename the temporary though, yes.